### ISSUE TITLE
gid/uid from userinfo ignored in v7.3.0 vs v6.8.0.

### ISSUE THREAD

**Issue by whataboutpereira** [OpenSMTPD/OpenSMTPD#1225]
My mailing list stopped working after upgrading from 6.8.0 to 7.3.0:

```
/usr/bin/mlmmj-receive[102515]: mlmmj-receive.c:122: Could not stat /var/vmail/mlmmj/domain/listname/: Permission denied
```

I changed permissions from 0700 vmail:vmail to 0750 vmail:smtpd and coaxed a more informative error out of mlmmj:

```
/usr/bin/mlmmj-receive[114552]: mlmmj-receive.c:112: Have to invoke either as root or as the user owning listdir Invoked with uid = [997]
```

So it seems OpenSMTPD is trying to deliver with uid 997 (which is 'smtpd' user on the system), ignoring the uid/git from userinfo table (5000:5000 correspond to vmail:vmail on the system).

v6.8.0:

```
lookup: lookup "3" as USERINFO in table proc:userinfo -> "5000:5000:/var/vmail/domain/user"
debug: smtpd: forking mda for session eba6bc20bdd7e6b8: 3 as 3

lookup: lookup "23" as USERINFO in table proc:userinfo -> "5000:5000:/var/vmail/mlmmj/domain/list"
debug: smtpd: forking mda for session 358aa0231bcacbe7: 23 as 23
```

v7.3.0:

```
lookup: lookup "3" as USERINFO in table proc:userinfo -> "5000:5000:/var/vmail/domain/user"
debug: smtpd: forking mda for session 8ed13ca11c1e1c09: 3 as smtpd

lookup: lookup "23" as USERINFO in table proc:userinfo -> "5000:5000:/var/vmail/mlmmj/domain/list"
debug: smtpd: forking mda for session 8ed13c9dff53f513: 23 as smtpd
```

Delivery to a regular user most likely succeeds because it's done over LMTP.

All userinfo is coming from table-mysql.

**Comment by whataboutpereira** [comment]
I've now found two ways to get mlmmj working.

.forward file method fails with the lmtp action that I've been successfully using thus far because it tries to pass it to mlmmj as 'smtpd' user:
```
action deliver_mlmmj    lmtp "/var/run/dovecot/lmtp" rcpt-to virtual <mlmmj_virtuals> userbase <mlmmj_userinfo>
```

Using the mda action passing mail to mlmmj works both with and without the .forward file, both ways delivered as the user from userbase:
```
action deliver_mlmmj   mda "/usr/bin/mlmmj-receive -L /var/vmail/mlmmj/%{rcpt.domain}/%{rcpt.user}/" virtual <mlmmj_virtuals> userbase <mlmmj_userinfo>
```

.forward file method also works if I replace lmtp with expand-only - mail is passed to mlmmj as the user from userbase not as 'smtpd' user:
```
action deliver_mlmmj   expand-only virtual <mlmmj_virtuals> userbase <mlmmj_userinfo>
```

**Comment by omar-polo** [comment]
Hello, and sorry for the delay.

This is caused by a changed made back in 2020: https://marc.info/?l=openbsd-tech&m=159024290713854&w=2

The rationale is to have fewer parts of smtpd requiring/running as root, so the lmtp delivery was changed to use the smtpd user instead of the recipient (switching users requires root permissions.)

It seems I got confused with the dates and haven't mentioned this change in the latest release ChangeLog, apologize.

Unfortunately I don't know mlmmj so I can't recommend a specific configuration change, I'm sorry.
